eng/pipelines/templates/steps/analyze.yml

```
@@ -0,0 +1,120 @@
+parameters:
+  - name: Artifacts
+    type: object
+    default: []
+  - name: ServiceDirectory
+    type: string
+    default: not-specified
+  - name: SDKType
+    type: string
+    default: all
+  - name: BuildSnippets
+    type: boolean
+    default: true
+
+steps:
+  - ${{ if not(contains(variables['Build.DefinitionName'], '-pr - ')) }}:
+    - template: /eng/common/pipelines/templates/steps/sparse-checkout.yml
+      parameters:
+          Paths:
+            - "/*"
+            - "!SessionRecords"
+
+  - template: /eng/pipelines/templates/steps/install-dotnet.yml
+
+  - template: /eng/common/pipelines/templates/steps/cache-ps-modules.yml
+
+  - template: /eng/common/pipelines/templates/steps/save-package-properties.yml
+    parameters:
+      ServiceDirectory: ${{ parameters.ServiceDirectory }}
+
+  - template: /eng/pipelines/templates/steps/set-artifact-packages.yml
+    parameters:
+      PackageInfo: '$(Build.ArtifactStagingDirectory)/PackageInfo'
+      Artifacts: ${{ parameters.Artifacts }}
+
+  - template: /eng/common/pipelines/templates/steps/verify-samples.yml
+    parameters:
+      ServiceDirectories: "$(ChangedServices)"
+
+  - template: /eng/common/pipelines/templates/steps/verify-readmes.yml
+    parameters:
+      PackagePropertiesFolder: '$(Build.ArtifactStagingDirectory)/PackageInfo'
+
+  - task: NodeTool@0
+    inputs:
+      versionSpec: 18.x
+    displayName: 'Use Node 18.x'
+
+  - task: DotNetCoreCLI@2
+    displayName: 'Install Snippet Generator Tool'
+    inputs:
+      command: custom
+      custom: 'tool'
+      arguments: 'restore'
+      workingDirectory: '$(Agent.BuildDirectory)'
+
+  - pwsh: |
+      $failed = $false
+      if ("$(ProjectNames)") {
```

```suggestion
      if ('$(ProjectNames)' -and '$(ProjectNames)' -notlike '$(*') {
```

Avoids the edge case where the variable remains unset.

---

Or access it via `$env:ProjectNames` which will be empty if not set.

---

yes but it would need to be `$env:PROJECTNAMES` since the linux env vars are case sensitive (or windows, I forget which)

---

Is this even required though? `$(ProjectNames)` will never be not set? Even if we resolve no packages (which isn't possible anymore due to Azure.Template fallback), it'll still be _populated_.

---

I took it, and also cleaned up a similar location for `build.yml`.

---

Yeah my suggestion was just defensive programming, because we have run into these bugs before when a variable later disappears, and it's even more annoying to debug when we don't halt the program as a result.